FogCreek/Glitch-Community
# Tall cheese: rebuild current user

## Links
* Remix link: https://tall-cheese.glitch.me/
* Issue-tracker link: https://app.zenhub.com/workspaces/community-5d07d1359dee6059a9688176/issues/fogcreek/glitch-community-work/352

## Changes:
* move superuser helpers into their own hook
* rebuild current user in redux

## How To Test:
* open two tabs to https://tall-cheese.glitch.me/
* sign into one of the tabs
* you should be signed into both tabs
* refresh one of the tabs
* it should first show an anonymous user, then your cached user state (showing a loader in place of your recent projects), then your fresh user state
* sign out of one of the tabs
* you should be signed out of both tabs

## Where am I going with this:
We've gotten pretty far just using context instead of Redux; what's the point of adding another library?
### Consistency
Our state is currently managed by a collection of interdependent context providers, each of which has its own state and interface. Most of our significant interactions require coordinating many of these at once: adding a user to a team project interacts with the api, the current user, notifications, error handlers, and the projects cache. Right now, the responsibility for wrangling all of this falls on the consumer, and this means that its easy to have two places that perform the same basic action but have inconsistent secondary effects (e.g. the notification is inconsistently displayed, the cache is not updated).

Redux encourages consistency both in how these systems are designed and in how consumers interact with them: the caller is responsible only for dispatching a message, and each system responds to that message by either transforming the state (as a reducer does) or by running a side effect (as a middleware does). This ensures that actions will have consistent behaviors, regardless of where they're dispatched from, and changing behavior (e.g. using a different API endpoint, changing the notification style) only needs to be done in one place.

### Performance
There are some significant performance "gotchas" when using "vanilla" React context -- context forces its consumer to re-render whenever its value changes, so you occasionally need to [split a context into multiple contexts for performance reasons](https://github.com/FogCreek/Glitch-Community/pull/683#discussion_r298367243), and you need to `useMemo` any functions you include to trigger actions. In practice, this means that we tend not to fix performance issues until they become painful.

Redux mitigates these issues by passing the _store_ as its context value, and setting up subscriptions on that; action creators are static and don't need to be memoized. There are still cases where one needs to optimize for performance in Redux, but in general it has better default behavior.
